collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java

```diff
@@ -0,0 +1,142 @@
+ * as a space-separated list of the response values. Most form builders either warn
+ * or error when a user tries to build a form with multiple selection questions
+ * that have values with spaces but it's possible to include them when writing
+ * XML by hand or using external choices.
+ *
+ * This class is responsible for checking if the underlying values in the set of answers
+ * contain spaces. If the values contain spaces, the warning is rendered in QuestionWidget.
+ *
+ * The class is composed of 2 ingredients:
+ * - SpacesInUnderlyingValues: check if the underlying values contain spaces
+ * - WarningRenderer: create and display the warning text
+ *
+ * Both classes are combined in the {@link #renderWarningIfNecessary(List)}
+ * that is responsible for applying the checker to values, and rendering warning if checker
+ * detects an issue.
+ *
+ * The class can be extended to render warnings into other widgets or in other way.
+ * To do that one should implement a new WarningRenderer, and create a new static
+ * initializer that initializes SpacesInUnderlyingValuesWarning with the new renderer
+ * as in {@link ##forQuestionWidget(QuestionWidget)}
+ *
+ * Created as fix for https://github.com/opendatakit/collect/issues/1964
+ *
+ */
+public class SpacesInUnderlyingValuesWarning {
+
+    private final SpacesInUnderlyingValues valuesChecker;
+    private final WarningRenderer warningRenderer;
+
+    public static SpacesInUnderlyingValuesWarning forQuestionWidget(QuestionWidget questionWidget) {
+        WarningRenderer renderer = new RenderIntoQuestionWidget(questionWidget);
+        SpacesInUnderlyingValues valuesChecker = new SpacesInUnderlyingValuesImpl();
+        return new SpacesInUnderlyingValuesWarning(valuesChecker, renderer);
+    }
+
+    @VisibleForTesting
+    SpacesInUnderlyingValuesWarning(SpacesInUnderlyingValues valuesChecker, WarningRenderer warningRenderer) {
+        this.valuesChecker = valuesChecker;
+        this.warningRenderer = warningRenderer;
+    }
+
+
+    public void renderWarningIfNecessary(List<SelectChoice> items) {
+        valuesChecker.check(items);
+
+        if (valuesChecker.hasInvalidValues()) {
+            warningRenderer.render(valuesChecker.getInvalidValues());
+        }
+    }
+
+    interface WarningRenderer {
+        void render(List<SelectChoice> items);
+    }
+
+    private static class RenderIntoQuestionWidget implements WarningRenderer {
+
+        private QuestionWidget questionWidget;
+        private final WarningTextCreator warningCreator;
+
+        RenderIntoQuestionWidget(QuestionWidget questionWidget) {
+            this.questionWidget = questionWidget;
+            warningCreator = new WarningTextCreatorImpl();
+        }
+
+        @Override
+        public void render(List<SelectChoice> invalidItems) {
+            questionWidget.showWarning(warningCreator.create(invalidItems, questionWidget.getContext()));
+        }
+    }
+
+    interface SpacesInUnderlyingValues {
```

Comment:
It feels too bad that the interface name is so specific. This really could be called something more generic like `UnderlyingValueChecker`, right? I also don't love using `Impl` as part of a class name so perhaps that could then become `SpacesUnderlyingValueChecker` or something like that.

Reply:
Yes, that looks much better, I'll apply the changes.